### PR TITLE
Request error handlers

### PR DESCRIPTION
**Pull request by pape-io**
You can provide custom error handlers to handle errors happening during the request.

If a error handler is provided nothing is raised.

If your error handler returns anything else but `nil` it replaces the response body.

```ruby
handler = ->{ do_something; return {name: 'unknown'} }
response = LHC.get('http://something', error_handler: handler)
response.data.name # 'unknown'
```
---------

- Required for some LHS feature to chain error handling and inject data in case of errors
- Minor version increase ahead

### CODE REVIEW

**Comment by Averethel** [comment]
@pape-io can we consider some naming unification? It's `error_hander` when used as option but `.handler` in chain...

File: lib/lhc/response.rb [hunk 1]
```diff
@@ -25,7 +25,7 @@ def effective_url
   end
 
   def body
-    raw.body
+    body_replacement || raw.body
```
Comment (Averethel) [inline lib/lhc/response.rb]: Why is this needed?

**Comment by Averethel** [comment]
Other than that small question 👍

**Comment by gonzalocasas** [comment]
Can't this be achieved more generally with interceptors instead of having explicit support for a specific kind of interception (i.e. error handling)?

**Comment by pape-io** [comment]
Good point @gonzalocasas. I will check into it 👍

**Comment by pape-io** [comment]
@Averethel please don't mix up LHS and LHC.

File: lib/lhc/response.rb (hunk 1 above)
Comment (pape-io) [inline lib/lhc/response.rb]: To set the response body with the error handler.

`raw.body` is immutable.

https://github.com/local-ch/lhc/pull/59/files#diff-d0e9c0bb6c4b2f9c1fae864c8323cbedR76
https://github.com/local-ch/lhc/pull/59/files#diff-5f35d7b3a1874ba18f7cf6077c5a20bcR106

**Comment by pape-io** [comment]
@gonzalocasas I'll checked if it's possible to add this functionality with an interceptor only. But it's not. 
error raising is an essential part of LHC... 
https://github.com/local-ch/lhc/pull/59/files#diff-5f35d7b3a1874ba18f7cf6077c5a20bcL100
An interceptor could not prevent LHC from raising an exception as Thyphoes responses are more or less immutable in this regard.

Do you think we should rather have an interface towards interceptors, and have the error handling logic in an interceptor iteself?

Would be glad to hear back @gonzalocasas

**Comment by gonzalocasas** [comment]
@pape-io I think delegating error handling to interceptors via an interface would improve composability, but if it complicates the architecture, then no need to go there.

**Comment by Averethel** [comment]
@pape-io why can't we have naming consistency between the two? Especially that one is a base for the other...

I'm with @gonzalocasas on the interceptor front but only if it's not overcomplicated.

**Comment by pape-io** [comment]
@Averethel It's consistent, it's just 2 different signatures to start with:

LHS
```ruby
Record.handle(LHC::Error, handler)
```

LHC
```ruby
LHC.get('http://local.ch', error_handler: handler)
```

The reason why I don't want to call it handler is, because the signature is not containing the type off Class it should handle. This additional (complexity) is part of the more complex library (LHS).

Does it make sense what I'm saying @Averethel ?

**Comment by pape-io** [comment]
@gonzalocasas @Averethel I checked again if it is possible to make an interceptor out of it. It's not.

The problem is that LHS needs some functionality that requires error handling to be part of LHC:
https://github.com/local-ch/lhs/pull/151

LHS can't or should add/manipulate/enable interceptors
https://github.com/local-ch/lhc/blob/master/docs/interceptors.md#global-default-interceptors
as this would happen implicit and is error-prone in regards of other things that might change the list of applied interceptors.

I understood your concern of composability but it seems error handling requires to be an essential part of LHC:
https://github.com/local-ch/lhc/blob/master/docs/exceptions.md

I hope it's okay we take the current solution, we might discover another solution in the future as both libraries evolve and I will keep it in mind @gonzalocasas
